bin/wasm-node/rust/src/lib.rs

```diff
@@ -413,45 +420,52 @@ async fn start_services(
 
     debug_assert!(per_chain.iter().all(Option::is_some));
 
-    // Spawn the JSON-RPC service. It is responsible for answer incoming JSON-RPC requests and
-    // sending back responses.
-    // Since multiple JSON-RPC services would conflict with each other, we start one for the first
-    // chain in the list.
-    let first_chain_services = per_chain.into_iter().next().unwrap().unwrap();
-    let first_chain_finalized_header = genesis_chain_information
-        .into_iter()
-        .next()
-        .unwrap()
-        .finalized_block_header;
-    let transactions_service = Arc::new(
-        transactions_service::TransactionsService::new(transactions_service::Config {
-            tasks_executor: Box::new({
-                let new_task_tx = new_task_tx.clone();
-                move |fut| new_task_tx.unbounded_send(fut).unwrap()
-            }),
-            network_service: (network_service.clone(), 0),
-            sync_service: first_chain_services.0.clone(),
-        })
-        .await,
-    );
-    new_task_tx
-        .unbounded_send(
-            json_rpc_service::start(json_rpc_service::Config {
+    // Spawn the JSON-RPC services. They are responsible for answering incoming JSON-RPC requests.
+    for (chain_index, (((services, json_rpc_running), genesis_chain_information), chain_spec)) in
+        per_chain
+            .into_iter()
+            .zip(json_rpc_running)
+            .zip(genesis_chain_information)
+            .zip(chain_specs)
+            .enumerate()
+    {
+        let (sync_service, runtime_service) = services.unwrap();
+
+        let finalized_header = genesis_chain_information.finalized_block_header;
+        let transactions_service = Arc::new(
+            transactions_service::TransactionsService::new(transactions_service::Config {
                 tasks_executor: Box::new({
                     let new_task_tx = new_task_tx.clone();
                     move |fut| new_task_tx.unbounded_send(fut).unwrap()
                 }),
                 network_service: (network_service.clone(), 0),
-                sync_service: first_chain_services.0,
-                transactions_service,
-                runtime_service: first_chain_services.1,
-                chain_spec: chain_specs.into_iter().next().unwrap(),
-                genesis_block_hash: first_chain_finalized_header.hash(),
-                genesis_block_state_root: first_chain_finalized_header.state_root,
+                sync_service: sync_service.clone(),
             })
-            .boxed(),
-        )
-        .unwrap();
+            .await,
+        );
+
+        if json_rpc_running {
```

Comment:
The `transactions_service` is more or less part of the JSON-RPC service and doesn't need to be started if `json_rpc_running` is false.
I suggest to just do `if !json_rpc_running { continue; }`